src/bin/cachepot-dist/build.rs

```diff
@@ -265,45 +263,109 @@ impl OverlayBuilder {
+            nix::unistd::ForkResult::Parent { child } => {
+                drop(stdout_pipe.1);
+                drop(stderr_pipe.1);
+                drop(compressed_outputs_pipe.1);
+
+                child
+            }
+            nix::unistd::ForkResult::Child => {
+                drop(stdout_pipe.0);
+                drop(stderr_pipe.0);
+                drop(compressed_outputs_pipe.0);
+
+                // We need to catch these here errors so they *DON'T* escape to
+                // the forked parent's context.
+                // Forking is a Bad Idea^TM anyway, see note above.
+                let do_work = move || -> Result<ExitStatus> {
+                    // Fetch uid/gid of the parent namespace user in order to
+                    // subsequently map to it in the child user namespace
+                    let uid = nix::unistd::Uid::current();
+                    let gid = nix::unistd::Uid::current();
+
+                    // We must call this from a main thread, hence why we forked
+                    // in the first place.
+                    nix::sched::unshare(nix::sched::CloneFlags::CLONE_NEWUSER)
+                        .context("Failed to enter a new user namespace")?;
+
                     // Now mounted filesystems will be automatically unmounted when this thread dies
                     // (and tmpfs filesystems will be completely destroyed)
                     nix::sched::unshare(nix::sched::CloneFlags::CLONE_NEWNS)
                         .context("Failed to enter a new Linux namespace")?;
+
+                    // Equivalent of `unshare --map-current-user` - we need that
+                    // to mount overlayfs and access files owned by the parent
+                    // namespace user
+                    std::fs::write("/proc/self/uid_map", format!("{} {} 1", uid, uid))
+                        .context("Failed writing to uid_map")?;
+                    std::fs::write("/proc/self/setgroups", "deny")
+                        .context("Failed writing to setgroups")?;
+                    std::fs::write("/proc/self/gid_map", format!("{} {} 1", gid, gid))
+                        .context("Failed writing to gid_map")?;
+
                     // Make sure that all future mount changes are private to this namespace
-                    // TODO: shouldn't need to add these annotations
-                    let source: Option<&str> = None;
-                    let fstype: Option<&str> = None;
-                    let data: Option<&str> = None;
-                    // Turn / into a 'slave', so it receives mounts from real root, but doesn't propogate back
+                    // Equivalent of `unshare --propagation private` (when used with `--mount`)
                     nix::mount::mount(
-                        source,
+                        Some("none"),
                         "/",
-                        fstype,
+                        Option::<&str>::None,
```

Comment:
Not sure I like this better

Reply:
I'm impartial, I just liked that there's a single `mount` call without any extra bindings above, so everything here reads easier like a list of syscalls performed one after another. Do you want me to revert this?

Reply:
I find it harder to read since I have to lookup each of those None value meanings, so a rever would be appreciated.

Reply:
Done: https://github.com/paritytech/cachepot/pull/128/commits/8424bb775bef237d578ead6d076c79d7daa50bf3